File: src/algorithms/binary-trees/node-depths/solution2.go
package main

type BinaryTree struct {
	Value int
	Left, Right *BinaryTree
}

// Average case: when the tree is balanced
// O(n) time | O(h) space - where n is the number of nodes in
// the Binary Tree and h is the height of the Binary Tree
func NodeDepths(root *BinaryTree) int {
	return nodeDepthsHelper(root, 0)
}

func nodeDepthsHelper(root *BinaryTree, depth int) int {
	if root == nil {
		return 0
	}
	return depth + nodeDepthsHelper(root.Left, depth+1) + nodeDepthsHelper(root.Right, depth+1)
}

File: src/algorithms/recursion/nth-fibonacci/solution2.ts
interface Cache {
  [key: number]: number;
}

// O(n) Time | O(n) Space
export function getNthFib(n: number, memoize: Cache = { 1: 0, 2: 1 }) {
  if (n in memoize) {
    return memoize[n];
  } else {
    memoize[n] = getNthFib(n - 1, memoize) + getNthFib(n - 2, memoize);
    return memoize[n];
  }
}
